_source/_posts/2021-06-18-native-java-comparison.adoc

```diff
@@ -0,0 +1,683 @@
+
+// The order of frameworks is intentional as I found Micronaut the easiest to make work and Spring Boot the most difficult.
+
+**Prerequisites**
+
+- https://sdkman.io/[Java 11]+
+- https://httpie.io/[HTTPie] (a better version of cURL)
+- An https://cloud.google.com/[Okta Developer] Account
+
+toc::[]
+
+== Get Started with Native Java Frameworks
+
+I created a GitHub repository you can clone and run to get started with all three frameworks quickly.
+
+[source,shell]
+----
+git clone https://github.com/oktadev/native-java-examples.git
+----
+
+This project has directories with the latest versions of Micronaut, Quarkus, and Spring Boot (at the time of this writing). I'll show you how I created them in individual sections below.
+
+Open the `native-java-examples` directory in your favorite IDE, so you have easy access to each framework's project files.
+
+TIP: If you want to see how to build native images in each framework, skip to the link:/blog/2021/06/18/native-java-comparison#build-native-images-for-micronaut-quarkus-and-spring-boot[build native images for Micronaut, Quarkus, and Spring Boot] section.
+
+== Launch a Micronaut Java API
+
+In a terminal window, cd into the `micronaut` directory and run `mn:run` to start it.
+
+[source,shell]
+----
+cd micronaut
+./mvn mn:run
+----
+
+If you open another terminal window and try to access it with HTTPie, you'll get a 401 Unauthorized error.
+
+[source,shell]
+----
+$ http :8080/hello
+
+HTTP/1.1 401 Unauthorized
+connection: keep-alive
+transfer-encoding: chunked
+----
+
+// To see them in action, you'll need to create an Okta developer account at https://developer.okta.com/signup[developer.okta.com/signup] or by running `okta register` with the https://cli.okta.com[Okta CLI].
+
+To make it so you can access this endpoint, you'll need to generate an OAuth 2.0 access token and update the JWKS (JSON Web Key Sets) URL to yours (in this project's `application.yml`).
+
+NOTE: If you're not sure what OIDC and OAuth 2.0 are, see our https://developer.okta.com/blog/2019/10/21/illustrated-guide-to-oauth-and-oidc[Illustrated Guide to OAuth and OpenID Connect].
+
+{% include setup/cli.md type="spa" loginRedirectUri="https://oidcdebugger.com/debug" logoutRedirectUri="https://oidcdebugger.com" %}
+
+Take note of the `clientId` and `issuer` values. You'll need those to get an access token and to configure each framework for JWT authentication.
+
+Open `micronaut/src/main/resources/application.yml` and change the Okta URL to match yours.
+
+[source,yaml]
+----
+micronaut:
+  application:
+    name: app
+  security:
+    enabled: true
+    token:
+      jwt:
+        enabled: true
+        signatures:
+          jwks:
+            okta:
+              url: https://{yourOktaDomain}/oauth2/default/v1/keys
+----
```

Comment:
I noticed you did issuer verification for the Quarkus and SB apps.  You can do that with Micronaut with:

```
micronaut:
  security:
    token:
      jwt:
        claims-validators:
          issuer: https://{yourOktaDomain}/oauth2/default
```

Reply:
Thanks, Jason! I'll give this a try.

Reply:
Do I still need all this configuration, or can I remove the JWKS URI?

```yaml
micronaut:
  application:
    name: app
  security:
    enabled: true
    token:
      jwt:
        enabled: true
        claims-validators:
          issuer: https://dev-1309757.okta.com/oauth2/default
        signatures:
          jwks:
            okta:
              url: https://dev-1309757.okta.com/oauth2/default/v1/keys
```

Reply:
@JasonTypesCodes Can you please answer this question? ^^

Reply:
Uggh, Micronaut's native app just stopped working for me. I thought it might be caused by the `claim-validators` line, so I removed it. However, it still happens. I'm on a different network than yesterday, but I don't think that should affect it.

If I run the app with `./mvnw mn:run` and hit it with an access token, it works. If I start the native app with  `./target/app`, it returns a 401. So strange...

Reply:
> Do I still need all this configuration, or can I remove the JWKS URI?

The issuer claim validator is separate from the JWKS config.  The configuration above looks correct

Reply:
> If I run the app with ./mvnw mn:run and hit it with an access token, it works. If I start the native app with ./target/app, it returns a 401. So strange...

I'd be very surprised if that broken native image.  Has anything else changed?

Reply:
Yeah, the 401 when away when I tested things again. A reboot might've helped. ;)

Can you confirm that the Micronaut issuer/JWKS look-up happens on the first request rather than at startup?

Reply:
> Can you confirm that the Micronaut issuer/JWKS look-up happens on the first request rather than at startup?

Singleton-scoped beans are intentionally lazy-loaded in Micronaut to improve startup performance and that includes the JWT Token Validator (which also includes the issuer claim validator)